ethereum_clients/PluginHost.js

```diff
@@ -1,34 +1,158 @@
 const fs = require('fs')
 const path = require('path')
+const { EventEmitter } = require('events')
 const { Plugin, PluginProxy } = require('./Plugin')
+const { getPluginCachePath } = require('./util')
+const { AppManager } = require('@philipplgh/electron-app-manager')
+
+function requireFromString(src, filename) {
+  var Module = module.constructor
+  var m = new Module()
+  m._compile(src, filename)
+  return m.exports
+}
 
 // TODO add file to electron packaged files
-class PluginHost {
+class PluginHost extends EventEmitter {
   constructor() {
+    super()
     this.plugins = []
     this.discover()
+      .then(plugins => {
+        this.plugins = this.plugins.concat(plugins)
+        // TODO emit plugins discovered
+      })
+      .catch(err => console.log('plugins could not be loaded', err))
+
+    this.discoverRemote()
+      .then(plugins => {
+        this.plugins = this.plugins.concat(plugins)
+        console.log(`${plugins.length} remote plugins found`)
+      })
+      .catch(err => {
+        console.log('remote plugins could not be loaded', err)
+      })
+      .finally(() => {
+        this.emit('plugins-loaded')
+      })
+  }
+  loadPluginFromFile(fullPath) {
+    const pluginConfig = require(fullPath)
+    // 2. TODO validate / verify
+    const plugin = new Plugin(pluginConfig)
+    return plugin
   }
-  // TODO should be async and not block UI init
-  discover() {
+  async loadPluginFromPackage(pluginManager, pkg) {
+    const index = await pluginManager.getEntry(pkg, 'package/index.js')
+    const indexContent = await index.file.readContent()
+    const pluginConfig = requireFromString(
+      indexContent.toString(),
+      `${pkg.location}/package/index.js`
+    )
+    const plugin = new Plugin(pluginConfig)
+    return plugin
+  }
+  async discoverRemote() {
+    const PLUGIN_DIR = path.join(__dirname, 'client_plugins')
+    let remotePluginList = []
+    try {
+      remotePluginList = JSON.parse(
+        fs.readFileSync(path.join(PLUGIN_DIR, 'plugins.json'))
+      )
+    } catch (error) {
+      console.log('error: could not parse plugin list', error)
+    }
+    let releases = remotePluginList.map(async pluginShortInfo => {
+      try {
+        const { name: pluginName, location } = pluginShortInfo
+        if (!location) {
+          throw new Error(
+            `error: external plugin ${pluginName} does not specify a valid location`
```

Comment:
I think we should perform a validation over whole plugin schema here, so we don't get unexpected results down the road.

Reply:
good catch. fix is provided through new app-manager version